TESTS/mbedmicro-rtos-mbed/signals/main.cpp

```diff
@@ -15,49 +15,348 @@
+
+    // 5.1    check clear on all flags with 0
+    ret = thread.signal_set(ALL_FLAGS);
+    TEST_ASSERT_EQUAL(ALL_FLAGS, ret);
+    WAIT1();
+
+    // 5.3
+    TEST_ASSERT_EQUAL(ALL_FLAGS, ret_val);
+    WAIT1();
+
+    // 5.5
+    TEST_ASSERT_EQUAL(ALL_FLAGS, ret_val);
+    WAIT1();
 
-DigitalOut led(LED1);
-int signal_counter = 0;
+    // 6.2    clear flags
+    TEST_ASSERT_EQUAL(ALL_FLAGS, ret_val);
+    WAIT1();
 
-void led_thread() {
-    while (true) {
-        // Signal flags that are reported as event are automatically cleared.
-        Thread::signal_wait(SIGNAL_SET_VALUE);
-        led = !led;
-        signal_counter++;
+    // 6.4
+    TEST_ASSERT_EQUAL(0, ret_val);
+
+    int32_t signals = 0;
+    for (int i = 0; i <= MAX_FLAG_POS; i++) {
+        int32_t signal = 1 << i;
+
+        // 7.1
+        ret = thread.signal_set(signal);
+        signals |= signal;
+        TEST_ASSERT_EQUAL(signals, ret);
+        WAIT1();
+
+        // 7.3
+        TEST_ASSERT_EQUAL(signals, ret);
     }
 }
 
-int main (void) {
-    GREENTEA_SETUP(20, "default_auto");
+/** Test signal_wait with no signals and 0 timeout
+
+    Given a thread
+    When wait for no signal specified for 0[ms] timeout
+    Then the @a signal_wait return with osOK
+ */
+void test_wait_0(void)
+{
+    Thread thread(osPriorityNormal, TEST_STACK_SIZE);
+    thread.start(callback(run_wait<0, 0>));
+    sync_sem1.wait();
+    TEST_ASSERT_EQUAL(osOK, ret_status);
+}
+
+/** Test signal_wait with signals and timeout
+
+    Given a thread
+    When wait for no signal specified for 1[ms] timeout
+    Then the @a signal_wait return with osOK
+    When wait for all signals for 1[ms] timeout
+    Then the @a signal_wait return with osEventTimeout
+ */
+void test_wait_timeout(void)
+{
+    {
+        Thread thread(osPriorityNormal, TEST_STACK_SIZE);
+        thread.start(callback(run_wait<0, 1>));
+        sync_sem1.wait();
+        TEST_ASSERT_EQUAL(osEventTimeout, ret_status);
+    }
+
+    {
+        Thread thread(osPriorityNormal, TEST_STACK_SIZE);
+        thread.start(callback(run_wait<ALL_FLAGS, 1>));
+        sync_sem1.wait();
+        TEST_ASSERT_EQUAL(osEventTimeout, ret_status);
+    }
+}
+
+/** Test signal_set (in no wait state) and signal_wait
+
+    Given a thread with no signal wait state
+    When call @a signal_set with specified signal
+    Then the it return resultant set signals
+    When thread reach signal wait state
+    Then it return with osEventSignal return code
+ */
+void wait_all_no_wait_state(void)
+{
+    int32_t ret;
+    Thread thread(osPriorityNormal, TEST_STACK_SIZE);
+    thread.start(callback(run_wait3<ALL_FLAGS, osWaitForever>));
+
+    sync_sem1.wait();
+    ret = thread.signal_set(ALL_FLAGS);
+    TEST_ASSERT_EQUAL(ALL_FLAGS, ret);
+    sync_sem2.release();
+
+    sync_sem1.wait();
+    TEST_ASSERT_EQUAL(osEventSignal, ret_status);
+}
+
+/** Test signal_set (in wait state) and signal_wait
 
+    Given a thread in signal wait state
```

Comment:
I thin that should be rewritten to closer match the test case, something like:
Given two threads A and B
When thread A waits for a signal
and thread B calls @a signal_set
Then thread A wakes up and receives correct signal